OpenRA.Mods.Common/Traits/Carryable.cs

```diff
@@ -111,18 +113,25 @@ public virtual void UnReserve(Actor self)
 		}
 
 		// Prepare for transport pickup
-		public virtual bool LockForPickup(Actor self, Actor carrier)
+		public virtual LockResponse LockForPickup(Actor self, Actor carrier)
 		{
-			if (state == State.Locked)
-				return false;
+			if (state == State.Locked && Carrier != carrier)
+				return LockResponse.Failed;
 
-			state = State.Locked;
-			Carrier = carrier;
+			if (state != State.Locked)
+			{
+				state = State.Locked;
+				Carrier = carrier;
 
-			if (lockedToken == ConditionManager.InvalidConditionToken && !string.IsNullOrEmpty(Info.LockedCondition))
-				lockedToken = conditionManager.GrantCondition(self, Info.LockedCondition);
+				if (lockedToken == ConditionManager.InvalidConditionToken && !string.IsNullOrEmpty(Info.LockedCondition))
+					lockedToken = conditionManager.GrantCondition(self, Info.LockedCondition);
+			}
 
-			return true;
+			var mobile = self.TraitOrDefault<Mobile>();
```

Comment:
Can we include the `// HACK` comment above here too?

Reply:
Even when we can trust `CurrentMovementTypes` we would probably want to check `FromCell` and `ToCell` anyway, because in that way we are also checking if we are actually on the grid.

Reply:
Can we put *that* in the comment then? :smile:

Reply:
I've created a new `IsMovingBetweenCells` property to make this more self-documenting and separated from the internal implementation of `Mobile`.

Reply:
Good idea. Can we please cache `mobile` here from `Created`, though?